modules/task/migrate.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 Gitea. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package task
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"runtime"
+	"strings"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/structs"
+	"code.gitea.io/gitea/modules/util"
+)
+
+func handleCreateError(owner *models.User, err error, name string) error {
+	switch {
+	case models.IsErrReachLimitOfRepo(err):
+		return fmt.Errorf("You have already reached your limit of %d repositories", owner.MaxCreationLimit())
+	case models.IsErrRepoAlreadyExist(err):
+		return errors.New("The repository name is already used")
+	case models.IsErrNameReserved(err):
+		return fmt.Errorf("The repository name '%s' is reserved", err.(models.ErrNameReserved).Name)
+	case models.IsErrNamePatternNotAllowed(err):
+		return fmt.Errorf("The pattern '%s' is not allowed in a repository name", err.(models.ErrNamePatternNotAllowed).Pattern)
+	default:
+		return err
+	}
+}
+
+func runMigrateTask(t *models.Task) error {
+	opts, err := t.MigrateConfig()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if e := recover(); e != nil {
+			var buf bytes.Buffer
+			fmt.Fprintf(&buf, "Handler crashed with error: %v", e)
+
+			for i := 1; ; i++ {
+				_, file, line, ok := runtime.Caller(i)
+				if !ok {
+					break
+				} else {
+					fmt.Fprintf(&buf, "\n")
+				}
+				fmt.Fprintf(&buf, "%v:%v", file, line)
+			}
+
+			err = errors.New(buf.String())
+		}
+
+		if err != nil {
+			t.EndTime = util.TimeStampNow()
+			t.Status = structs.TaskStatusFailed
+			t.Errors = err.Error()
+			if err := t.UpdateCols("status", "errors", "end_time"); err != nil {
+				log.Error("Task UpdateCols failed: %s", err.Error())
+			} else if t.Repo != nil {
+				if errDelete := models.DeleteRepository(t.Doer, t.OwnerID, t.Repo.ID); errDelete != nil {
+					log.Error("DeleteRepository: %v", errDelete)
+				}
+			}
+		} else {
+			if err := models.FinishMigrateTask(t); err != nil {
+				log.Error("Task UpdateCols failed: %s", err.Error())
+			}
+		}
+	}()
+
+	if err := t.LoadRepo(); err != nil {
+		return err
+	}
+
+	if err := t.LoadDoer(); err != nil {
+		return err
+	}
+	if err := t.LoadOwner(); err != nil {
+		return err
+	}
+	t.StartTime = util.TimeStampNow()
+	t.Status = structs.TaskStatusRunning
+	if err := t.UpdateCols("start_time", "status"); err != nil {
+		return err
+	}
+
+	repo, err := models.MigrateRepositoryGitData(t.Doer, t.Owner, t.Repo, *opts)
```

Comment:
I think you probably should check that the repo is still `Creating` (or `Initializing`/`Instantiating`) either here or in `MigrateRepositoryGitData` - and either fail if not or put it into the IsCreating status before you start doing anything.